Lib/test/test_heapq.py

```diff
@@ -0,0 +1,456 @@
+
+    def test_merge_does_not_suppress_index_error(self):
+        # Issue 19018: Heapq.merge suppresses IndexError from user generator
+        def iterable():
+            s = list(range(10))
+            for i in range(20):
+                yield s[i]       # IndexError when i > 10
+        with self.assertRaises(IndexError):
+            list(self.module.merge(iterable(), iterable()))
+
+    # TODO: RUSTPYTHON
+    # 'method_descriptor' object has no attribute '__self__'"
+    @unittest.expectedFailure
+    def test_merge_stability(self):
+        class Int(int):
+            pass
+        inputs = [[], [], [], []]
+        for i in range(20000):
+            stream = random.randrange(4)
+            x = random.randrange(500)
+            obj = Int(x)
+            obj.pair = (x, stream)
+            inputs[stream].append(obj)
+        for stream in inputs:
+            stream.sort()
+        result = [i.pair for i in self.module.merge(*inputs)]
+        self.assertEqual(result, sorted(result))
+
+    def test_nsmallest(self):
+        data = [(random.randrange(2000), i) for i in range(1000)]
+        for f in (None, lambda x:  x[0] * 547 % 2000):
+            for n in (0, 1, 2, 10, 100, 400, 999, 1000, 1100):
+                self.assertEqual(list(self.module.nsmallest(n, data)),
+                                 sorted(data)[:n])
+                self.assertEqual(list(self.module.nsmallest(n, data, key=f)),
+                                 sorted(data, key=f)[:n])
+
+    def test_nlargest(self):
+        data = [(random.randrange(2000), i) for i in range(1000)]
+        for f in (None, lambda x:  x[0] * 547 % 2000):
+            for n in (0, 1, 2, 10, 100, 400, 999, 1000, 1100):
+                self.assertEqual(list(self.module.nlargest(n, data)),
+                                 sorted(data, reverse=True)[:n])
+                self.assertEqual(list(self.module.nlargest(n, data, key=f)),
+                                 sorted(data, key=f, reverse=True)[:n])
+
+    def test_comparison_operator(self):
+        # Issue 3051: Make sure heapq works with both __lt__
+        # For python 3.0, __le__ alone is not enough
+        def hsort(data, comp):
+            data = [comp(x) for x in data]
+            self.module.heapify(data)
+            return [self.module.heappop(data).x for i in range(len(data))]
+        class LT:
+            def __init__(self, x):
+                self.x = x
+            def __lt__(self, other):
+                return self.x > other.x
+        class LE:
+            def __init__(self, x):
+                self.x = x
+            def __le__(self, other):
+                return self.x >= other.x
+        data = [random.random() for i in range(100)]
+        target = sorted(data, reverse=True)
+        self.assertEqual(hsort(data, LT), target)
+        self.assertRaises(TypeError, data, LE)
+
+
+class TestHeapPython(TestHeap, TestCase):
+    module = py_heapq
+
+
+@unittest.skip("TODO: RUSTPYTHON, _heapq module")
```

Comment:
Note: Could change these to `@skipUnless(c_heapq, 'requires _heapq')` as they were in the original source file.

Reply:
Yeah, I think that'd be good, if it's possible to leave something as it is it's usually good to do so.